src/Numerics/Mesh/Grids.jl

```diff
@@ -570,106 +591,242 @@ function commmapping(N, commelems, commfaces, nabrtocomm)
+    nelem = size(elemtocoord, 3)
+
     Nq = ntuple(j -> size(D[j], 1), dim)
     Np = prod(Nq)
     Nfp = div.(Np, Nq)
 
-    FT = eltype(D[1])
+    Nq_N1 = max.(Nq, 2)
+    Np_N1 = prod(Nq_N1)
+    Nfp_N1 = div.(Np_N1, Nq_N1)
+
+    # First we compute the geometry with all the N = 0 dimension set to N = 1
+    ξ1, ω1 = Elements.lglpoints(FT, 1)
+    D1 = Elements.spectralderivative(ξ1)
+    D_N1 = ntuple(j -> Nq[j] == 1 ? D1 : D[j], dim)
+    ξ_N1 = ntuple(j -> Nq[j] == 1 ? ξ1 : ξ[j], dim)
+    ω_N1 = ntuple(j -> Nq[j] == 1 ? ω1 : ω[j], dim)
+    (vgeo_N1, sgeo_N1, x_vtk) =
+        computegeometry(elemtocoord, D_N1, ξ_N1, ω_N1, meshwarp)
+
+
+    # Sort out the vgeo terms
+    @views begin
+        vgeo_N1_flds =
+            ntuple(fld -> reshape(vgeo_N1[:, fld, :], Nq_N1..., nelem), _nvgeo)
+
+        # Allocate the storage for N = 0 volume metrics
+        vgeo = zeros(FT, Np, _nvgeo, nelem)
+
+        # Counter to make sure we got all the vgeo terms
+        num_vgeo_handled = 0
+
+        # _M should be a sum
+        vgeo[:, _M, :][:] .= sum(vgeo_N1_flds[_M], dims = findall(Nq .== 1))[:]
+        num_vgeo_handled += 1
+
+        # need to recompute _MI
+        vgeo[:, _MI, :] = 1 ./ vgeo[:, _M, :]
+        num_vgeo_handled += 1
+
+        # coordinates should just be averages
+        avg_den = 2 .^ sum(Nq .== 1)
+        for fld in (_x1, _x2, _x3)
+            vgeo[:, fld, :] =
+                sum(vgeo_N1_flds[fld], dims = findall(Nq .== 1))[:] / avg_den
+            num_vgeo_handled += 1
+        end
+
+        # For the metrics it is J * ξixk we approximate so multiply and divide the
+        # mass matrix (which has the Jacobian determinant and the proper averaging
+        # due to the quadrature weights)
+        M_N1 = vgeo_N1_flds[_M]
+        MI = vgeo[:, _MI, :]
+        for fld in
+            (_ξ1x1, _ξ2x1, _ξ3x1, _ξ1x2, _ξ2x2, _ξ3x2, _ξ1x3, _ξ2x3, _ξ3x3)
+            vgeo[:, fld, :] =
+                sum(M_N1 .* vgeo_N1_flds[fld], dims = findall(Nq .== 1))[:] .*
+                MI[:]
+            num_vgeo_handled += 1
+        end
+
+        # compute MH and JvC
+        horizontal_vertical_metrics(vgeo, Nq, ω)
+        num_vgeo_handled += 2
+
+        # Make sure we handled all the vgeo terms
+        @assert _nvgeo == num_vgeo_handled
+    end
+
+    # Sort out the sgeo terms
+    @views begin
+        sgeo = zeros(FT, _nsgeo, maximum(Nfp), nface, nelem)
+
+        # for the volume inverse mass matrix
+        MI = vgeo[:, _MI, :]
+        p = reshape(1:Np, Nq)
+        if dim == 1
+            fmask = (p[1:1], p[Nq[1]:Nq[1]])
+        elseif dim == 2
+            fmask = (p[1, :][:], p[Nq[1], :][:], p[:, 1][:], p[:, Nq[2]][:])
+        elseif dim == 3
+            fmask = (
+                p[1, :, :][:],
+                p[Nq[1], :, :][:],
+                p[:, 1, :][:],
+                p[:, Nq[2], :][:],
+                p[:, :, 1][:],
+                p[:, :, Nq[3]][:],
+            )
+        end
```

Comment:
are these fmask, ordered as x direction flux faces, y direction flux faces, z direction flux faces? For example, I have N = [2;3;0], I have fmask ([1, 4, 7, 10], [3, 6, 9, 12], [1, 2, 3], [10, 11, 12], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]), what does this mean? is fmask used to map the face points to the volume points to correct MI?

Reply:
`fmask` (face mask) is to select the volume points associated with each face in the reference element. In this case it is being used to pull out the volume mass matrix on the face (which is necessary for the "lift" in DG

The faces are numbered in z-ordering, which I think is what you have

- face 1: `ξ1 = -1`
- face 2: `ξ1 = +1`
- face 3: `ξ2 = -1`
- face 4: `ξ2 = +1`
- face 5: `ξ3 = -1`
- face 6: `ξ3 = +1`